sqlx-core/src/postgres/types/decimal.rs

```diff
@@ -0,0 +1,420 @@
+        };
+
+        // weight is 0 if the decimal point falls after the first base-10000 digit
+        let scale = (digits.len() as i64 - weight as i64 - 1) * 4;
+
+        // no optimized algorithm for base-10 so use base-100 for faster processing
+        let mut cents = Vec::with_capacity(digits.len() * 2);
+        for digit in &digits {
+            cents.push((digit / 100) as u8);
+            cents.push((digit % 100) as u8);
+        }
+
+        let bigint = BigInt::from_radix_be(sign, &cents, 100)
+            .ok_or("PgNumeric contained an out-of-range digit")?;
+
+        match (bigint.to_i128(), scale) {
+            // A negative scale, meaning we have nothing on the right and must
+            // add zeroes to the left.
+            (Some(num), scale) if scale < 0 => Ok(Decimal::from_i128_with_scale(
+                num * 10i128.pow(scale.abs() as u32),
+                0,
+            )),
+            // A positive scale, so we have decimals on the right.
+            (Some(num), _) => Ok(Decimal::from_i128_with_scale(num, scale as u32)),
+            (None, _) => Err("Decimal's integer part out of range.".into()),
+        }
+    }
+}
+
+impl TryFrom<&'_ Decimal> for PgNumeric {
+    type Error = BoxDynError;
+
+    fn try_from(decimal: &Decimal) -> Result<Self, BoxDynError> {
+        if decimal.is_zero() {
+            return Ok(PgNumeric::Number {
+                sign: PgNumericSign::Positive,
+                scale: 0,
+                weight: 0,
+                digits: vec![],
+            });
+        }
+
+        let scale = decimal.scale() as u16;
+
+        // A serialized version of the decimal number. The resulting byte array
+        // will have the following representation:
+        //
+        // Bytes 1-4: flags
+        // Bytes 5-8: lo portion of m
+        // Bytes 9-12: mid portion of m
+        // Bytes 13-16: high portion of m
+        let s = decimal.serialize();
+
+        // Moving the flags from the beginning of the array to the end, giving
+        // us a representation of u96 we can convert to u128.
+        //
+        // We also just set the flags as zero, so we don't need to chop them off
+        // from the resulting integer.
+        let mut mantissa = u128::from_le_bytes([
+            s[4], s[5], s[6], s[7], // lo portion
+            s[8], s[9], s[10], s[11], // mid portion
+            s[12], s[13], s[14], s[15], // hi portion
+            0, 0, 0, 0, // flags (cleared)
+        ]);
+
+        // If our scale is not a multiple of 4, we need to go to the next
+        // multiple.
+        let groups_diff = scale % 4;
+        if groups_diff > 0 {
+            let remainder = 4 - groups_diff as u32;
+            let power = 10u32.pow(remainder as u32) as u128;
+
+            mantissa = mantissa * power;
+        }
+
+        // Array to store max mantissa of Decimal in Postgres decimal format.
+        let mut digits = Vec::with_capacity(8);
```

Comment:
Hmm, maybe we should optimize `PgNumeric` with `SmallVec`. That's orthogonal to this though.

Reply:
Done.

Reply:
I tried to see how it would work and would like to leave it to another PR.